src/miner.erl

```diff
@@ -600,9 +445,232 @@ terminate(Reason, _State) ->
+        LastBlockTime -> LastBlockTime + 1;
+        NewTime       -> NewTime
+    end.
+
+-spec select_transactions(
+    blockchain:blockchain(),
+    blockchain_txn:txns(),
+    blockchain_block:block(),
+    non_neg_integer(),
+    non_neg_integer()
+) ->
+    {
+        ElectionEpoch :: non_neg_integer(),
+        EpochStart :: non_neg_integer(),
+        TxsValid   :: blockchain_txn:txns(),
+        TxsInvalid :: blockchain_txn:txns()
+    }.
+select_transactions(Chain, Txns, BlockCurr, BlockHeightCurr, BlockHeightNext) ->
+    {ElectionEpoch0, EpochStart0} = blockchain_block_v1:election_info(BlockCurr),
+    SortedTransactions =
+        lists:sort(fun blockchain_txn:sort/2, [T || T <- Txns, not txn_is_rewards(T)]),
+    {ValidTransactions, InvalidTransactions0} = blockchain_txn:validate(SortedTransactions, Chain),
+    %% InvalidTransactions0 is a list of tuples in the format {Txn, InvalidReason}
+    %% we dont need the invalid reason here so need to remove the tuple format
+    %% and have a regular list of txn items in prep for returning to hbbft.
+    InvalidTransactions = [InvTxn || {InvTxn, _InvalidReason} <- InvalidTransactions0],
+    Ledger = blockchain:ledger(Chain),
+    case blockchain_election:has_new_group(ValidTransactions) of
+            {true, _, ConsensusGroupTxn, _} ->
+                Epoch = ElectionEpoch0 + 1,
+                Start = EpochStart0 + 1,
+                End = BlockHeightCurr,
+                RewardsMod =
+                    case blockchain:config(?rewards_txn_version, Ledger) of
+                         {ok, 2} -> blockchain_txn_rewards_v2;
+                         _       -> blockchain_txn_rewards_v1
+                     end,
+                {ok, Rewards} = RewardsMod:calculate_rewards(Start, End, Chain),
+                lager:debug("RewardsMod: ~p, Rewards: ~p~n", [RewardsMod, Rewards]),
+                RewardsTxn = RewardsMod:new(Start, End, Rewards),
+                %% To cut down on the size of group txn blocks, which we'll
+                %% need to fetch and store all of to validate snapshots, we
+                %% discard all other txns for this block.
+                Transactions =
+                    lists:sort(
+                        %% TODO Rename blockchain_txn:sort to blockchain_txn:(compare|cmp)
+                        fun blockchain_txn:sort/2,
+                        [RewardsTxn, ConsensusGroupTxn]
+                     ),
+                {Epoch, BlockHeightNext, Transactions, InvalidTransactions};
+            _ ->
+                {ElectionEpoch0, EpochStart0, ValidTransactions, InvalidTransactions}
+        end.
+
+-spec txn_is_rewards(blockchain_txn:txn()) -> boolean().
+txn_is_rewards(Txn) ->
+    Rewards = [blockchain_txn_rewards_v1, blockchain_txn_rewards_v2],
+    lists:member(blockchain_txn:type(Txn), Rewards).
+
+-spec metadata_to_seen_bbas(metadata()) ->
+    [{{pos_integer(), binary()}, binary()}].
+metadata_to_seen_bbas(M) ->
```

Comment:
I'm not sure this passes the 'useful abstraction' test

Reply:
It was one of the first things to get plucked out of create_block as I was desperately trying to slim it down. It itself wasn't slim either... After several passes - it now looks this slim and can probably be put back in.

Reply:
Removed.